### ISSUE TITLE
Dockerfile will not build

### ISSUE THREAD

**Issue by nt3rp** [SoTrxII/Pandora#10]
**Problem**
When trying to run the Pandora application, the Docker image cannot be built.

**Steps to Reproduce**
1. `git clone` the latest release [`v2.5.2`](https://github.com/SoTrxII/Pandora/releases/tag/2.5.2)
    - The problem is also present in the latest version of [`master`](https://github.com/SoTrxII/Pandora/commit/7b0a21dd4d70981b5e0736df2ab83e52c98774cc)
3. `cp .env.example` to `samples/minimal` and rename to `.env`
4. Add your Discord bot token as `PANDORA_TOKEN`
5. `cd samples/minimal; docker compose up`

**Expected**
Pandora, along with a redis instance and cooking instance should be running.

**Actual**
An error occurs while trying to `yarn install`:

<details>
  <summary>Error</summary>
  
  ```
  docker compose up
  [+] Building 0.0s (0/1)                                                                                     
  [+] Building 0.2s (2/3)                                                                                                                                                                
  docker:default
     => [pandora internal] load build definition from Dockerfile                                                                                                                                     
     => => transferring dockerfile: 933B                                                                        
  [+] Building 0.4s (2/3)                                                                                                                                                                
  docker:default                   
     => [pandora internal] load build definition from Dockerfile                                                                                                                                     
     => => transferring dockerfile: 933B                                                                        
  [+] Building 19.9s (11/14)                                                                   
  docker:default
     => [pandora internal] load build definition from Dockerfile                                           
     => => transferring dockerfile: 933B                                                                   
     => [pandora internal] load metadata for docker.io/library/node:18-alpine                              
     => [pandora auth] library/node:pull token for registry-1.docker.io                                    
     => [pandora internal] load .dockerignore                                                              
     => => transferring context: 116B                                                                      
     => [pandora internal] load build context                                                              
     => => transferring context: 7.35kB                                                                    
     => [pandora build 1/7] FROM docker.io/library/node:18-alpine@sha256:7e43a2d633d91e8655a6c0f45d2ed987  
     => CACHED [pandora build 2/7] WORKDIR /app                                                            
     => CACHED [pandora build 3/7] RUN apk add python3 make alpine-sdk yarn                                
     => CACHED [pandora build 4/7] COPY . /app/                                                            
     => CACHED [pandora build 5/7] RUN yarn set version berry && grep -qF 'nodeLinker' .yarnrc.yml  || ec  
     => ERROR [pandora build 6/7] RUN yarn install                                                        
    ------                                                                                                      
     > [pandora build 6/7] RUN yarn install:                                                                    
    0.587 ➤ YN0087: Migrated your project to the latest Yarn version 🚀                                         
    0.587                                                                                                       
    0.589 ➤ YN0000: · Yarn 4.5.3                                                                                
    0.598 ➤ YN0000: ┌ Resolution step                                                                           
    4.408 ➤ YN0085: │ + 
  typescript@patch:typescript@npm%3A5.0.3#optional!builtin<compat/typescript>::version=5.0.3&hash=b5f058, and 2 more.
    4.408 ➤ YN0085: │ - fsevents@patch:fsevents@npm%3A2.3.2#~builtin<compat/fsevents>::version=2.3.2&hash=18f3a7, and 2 more.
    4.419 ➤ YN0000: └ Completed in 3s 821ms
    4.420 ➤ YN0000: ┌ Post-resolution validation
    4.420 ➤ YN0086: │ Some peer dependencies are incorrectly met by dependencies; run yarn explain peer-requirements for details.
    4.421 ➤ YN0000: └ Completed
    4.421 ➤ YN0000: ┌ Fetch step
    12.39 ➤ YN0013: │ 663 packages were added to the project (+ 41.06 MiB).
    12.39 ➤ YN0000: └ Completed in 7s 966ms
    12.42 ➤ YN0000: ┌ Link step
    16.59 ➤ YN0007: │ @discordjs/opus@npm:0.9.0 must be built because it never has been before or the last one failed
    16.59 ➤ YN0007: │ bufferutil@npm:4.0.7 must be built because it never has been before or the last one failed
    16.59 ➤ YN0007: │ zlib-sync@npm:0.1.8 must be built because it never has been before or the last one failed
    16.59 ➤ YN0007: │ restana@npm:4.9.4 must be built because it never has been before or the last one failed
    16.59 ➤ YN0007: │ protobufjs@npm:6.11.2 must be built because it never has been before or the last one failed
    18.47 ➤ YN0009: │ zlib-sync@npm:0.1.8 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-5830599b/build.log)
    18.88 ➤ YN0009: │ @discordjs/opus@npm:0.9.0 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-40ff01b3/build.log)
    18.88 ➤ YN0000: └ Completed in 6s 463ms
    18.91 ➤ YN0000: · Failed with errors in 18s 325ms
    ------
    failed to solve: process "/bin/sh -c yarn install" did not complete successfully: exit code: 1
    ```
</details>

**Comment by nt3rp** [comment]
I've done a bit of investigation, and have managed to figure out a few pieces but not the whole problem:

**`node:18-alpine` doesn't ship with `disttools`**
The image used for the Dockerfile ships with python3, which doesn't include `disttools`, which is used by `gyp`, which is a dependency of `zlib-sync` and `opus`. In my testing, fixing that was a matter of adding `py-setuptools` to lines 5 and 15 of [the Dockerfile](https://github.com/SoTrxII/Pandora/blob/7b0a21dd4d70981b5e0736df2ab83e52c98774cc/Dockerfile#L5)

**Incorrect version of `yarn` used**
After getting past that, I was running into issues with `yarn`. I don't have a log handy, but it was something like, "`packageManager` requires yarn@4.x (from `package.json`) but currently using 1.22.22" (which I was able to confirm).

I _think_ the solution to that is to run `corepack enable` before installing `yarn`, as it seemed to resolve the issue.

**Problems with `yarn`**
After that I hit a roadblock that I haven't figured out as I'm not super familiar with `yarn`. It's either:
- There's an issue setting the yarn version; or
- There's a problem installing the `workspaces` plugin

(I'll try to dig in deeper but wanted to file this bug report in case there are known fixes—or if someone more familiar with the project is able to fix it faster)

**Comment by SoTrxII** [comment]
Hey,  
Thank you for your issue!  
I just pushed a fix based on your feedback. As you pointed out, the `disttools` module required by `node-gyp` is no longer included in Python 3.12 and above, which is the default version in newer Alpine releases.

The Yarn version used in the Dockerfile was "Yarn Berry," the beta version of Yarn 2.0. Starting from Yarn 4.0, the workspace plugin is bundled with Yarn.

With these updates, you should now be able to build the image without any issues.

That said, this Dockerfile is quite outdated. I plan to update the Node.js version sometime in the future.

**Comment by ndcrd** [comment]
Hello! For some reason current fix of docker file didn't work for me. It couldn't build discordjs/opus. 
But I've came up with workaround to make a bit different dockerfile with pnpm instead of yarn to make it work

just want to share with someone who was stuck just as me.

```
# Build stage
FROM node:18-bookworm as build
WORKDIR /app

# Install required packages including python3-distutils and ffmpeg
RUN apt-get update && apt-get install -y \
    python3-distutils \
    ffmpeg \
    make \
    g++

# Install pnpm
RUN npm install -g pnpm

# Copy source files
COPY . /app/

# Install dependencies and build
RUN pnpm install
RUN pnpm run build

# Production stage
FROM node:18-bookworm as prod
WORKDIR /app

# Install runtime dependencies
RUN apt-get update && apt-get install -y \
    python3-distutils \
    ffmpeg \
    && rm -rf /var/lib/apt/lists/*

# Install pnpm
RUN npm install -g pnpm
# Copy built files from build stage
COPY --from=build /app/dist /app
COPY package.json /app/

# Install production dependencies only
RUN pnpm install --prod

# Run the application
CMD ["node", "/app/main.js"]
```
